docker/pandoc-lambda/function/app.py

```diff
@@ -1,14 +1,133 @@
+    doc_insert.attrib[custom_mark_att] = "1"
+    doc_insert.attrib[id_att] = id
+
+    # Content
+    for t in texts:
+        footnote.append(t)
+
+    # Insert into the footnotes file
+    footnotes_part.element.append(footnote)
+
+    # Insert the reference into the doc
+    ref.insert(1,doc_insert)
+
+def promote_case_footnotes(doc):
+    val_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val'
+    author_footnotes = {}
+
+    for ref in doc.element.xpath("//*[*/w:rStyle[starts-with(@w:val,'FootnoteReference')]]"):
+        mark_text = ref.text
+        style_node = ref.xpath(".//w:rStyle[starts-with(@w:val,'FootnoteReference')]")[0]
+        node_id = style_node.attrib[val_att][18:]
+        mark_id = f"{node_id}-{mark_text}"
+        if mark_id not in author_footnotes:
+            author_footnotes[mark_id] = {'id':mark_id, 'mark': mark_text, 'refs': [], 'texts': []}
+        style_node.attrib[val_att] = 'FootnoteReference'
+        author_footnotes[mark_id]['refs'].append(ref)
+
+    for text_el in doc.element.xpath("//w:p[w:pPr/w:pStyle[starts-with(@w:val,'FootnoteText')]]"):
+        mark_text = text_el.xpath(".//*[*/w:rStyle]/w:t")[0].text
+        node_id = text_el.xpath(".//w:pStyle[starts-with(@w:val, 'FootnoteText')]/@w:val")[0][13:]
+        mark_id = f'{node_id}-{mark_text}'
+        if mark_id not in author_footnotes:
+            author_footnotes[mark_id] = {'id':mark_id, 'mark': mark_text, 'refs': [], 'texts': []}
+        mark_el = text_el.xpath("./w:pPr/w:pStyle[starts-with(@w:val,'FootnoteText')]")[0]
+        mark_el.attrib[val_att] = 'FootnoteText'
+        for style in text_el.xpath(".//w:pStyle[starts-with(@w:val, 'FootnoteText')]"):
+            style.attrib[val_att] = 'FootnoteText'
+        for style in text_el.xpath(".//w:rStyle[starts-with(@w:val, 'FootnoteRef')]"):
+            style.attrib[val_att] = 'FootnoteReference'
+        author_footnotes[mark_id]['texts'].append([text_el])
+
+
+    case_footnotes = {}
+
+    for ref in doc.element.xpath("//*[*/w:rStyle[starts-with(@w:val,'CaseFootnoteReference')]]"):
+        mark_text = ref.text
+        node = ref.xpath(".//w:rStyle[starts-with(@w:val,'CaseFootnoteReference')]")[0]
+        node_id = node.attrib[val_att][22:]
+        mark_id = f'{node_id}-{mark_text}'
+        if mark_id not in case_footnotes:
+            case_footnotes[mark_id] = {'id':mark_id, 'mark': mark_text, 'refs': [], 'texts': []}
+        node.attrib[val_att] = "FootnoteReference"
+        parent = ref.getparent()
+        gp = parent.getparent()
+        gp.replace(parent, ref)
+        case_footnotes[mark_id]['refs'].append(ref)
+
+    for footnote_start in doc.element.xpath("//*[*/*[starts-with(@w:val,'CaseFootnoteText')] and .//w:hyperlink]"):
+        mark_text = footnote_start.xpath(".//*[*/w:rStyle]/w:t")[0].text
+        node_id = footnote_start.xpath(".//w:pStyle[starts-with(@w:val, 'CaseFootnoteText')]/@w:val")[0][17:]
+
```

Comment:
Testing my understanding: is the `current_stack` and corresponding `while` stanza to handle footnotes with multiple paragraphs of text?

I think this stanza might benefit from a comment stating its goal, since it's a little dense to read through.

Reply:
You're right, and I'll add some commentary for future perusers.